### PR TITLE
Linux sl 5.5 amd

### PR DESCRIPTION
**Pull request by krystian-hebel**

### CODE REVIEW

File: arch/x86/boot/compressed/sl_main.c [hunk 1]
```diff
@@ -278,30 +284,33 @@ void sl_main(u8 *bootparams)
...
+		if (sl_cpu_type == SL_CPU_INTEL)
+			sl_txt_reset(TXT_SLERROR_TPM_INIT);
+		else
+			sl_skinit_reset();
+	}
 
 	/* Locate the TPM event log. */
 	sl_find_event_log(tpm);
 
-	/*
```
Comment (rossphilipson) [inline arch/x86/boot/compressed/sl_main.c:299]: Why did you remove this comment? Daniel asked me to put it there IIRC

File: arch/x86/boot/compressed/sl_stub.S [hunk 2]
```diff
@@ -168,9 +168,27 @@ SYM_FUNC_START(sl_stub)
...
+	jnz	.Ldo_unknown_cpu
+
+	/* Know it is AMD */
+	movl	$(SL_CPU_AMD), sl_cpu_type(%ebp)
+
+	/*
+	 * Disable maskable interrups in EFLAGS then enable global interrupts
+	 * including SMI and NMI (GIF).
+	 */
+	cli
```
Comment (rossphilipson) [inline arch/x86/boot/compressed/sl_stub.S:189]: Maskable interrupts were already disabled above with the direction flag

File: arch/x86/boot/compressed/sl_main.c (hunk 1 above)
Comment (dpsmith) [reply arch/x86/boot/compressed/sl_main.c:299]: Yes please leave the comment so it is clear why locality 2 is opened vs. locality 1

File: arch/x86/boot/compressed/sl_main.c [hunk 3]
```diff
@@ -160,7 +166,7 @@ static void sl_tpm12_log_event(u32 pcr, u8 *digest,
 	total_size = sizeof(struct tpm12_pcr_event) + event_size;
 
 	if (tpm12_log_event(evtlog_base, total_size, pcr_event))
-		sl_txt_reset(SL_ERROR_TPM_LOGGING_FAILED);
+		sl_txt_reset(SL_ERROR_TPM_EXTEND);
```
Comment (dpsmith) [inline arch/x86/boot/compressed/sl_main.c:163]: Please constrain changes to adding AMD.

File: arch/x86/boot/compressed/tpm/tpm2_cmds.c [hunk 4]
```diff
@@ -84,11 +84,13 @@ int tpm2_extend_pcr(struct tpm *t, u32 pcr,
 	u16 size;
 	int ret = 0;
 
+	b = alloc_tpmbuff(t->intf, TPM_NO_LOCALITY);
```
Comment (dpsmith) [inline arch/x86/boot/compressed/tpm/tpm2_cmds.c:87]: IMHO we should not be allocating the buffer in the middle of the extend, it should have already been allocated. Though we should be checking that the buffer was allocated, i.e. != NULL.

Comment (rossphilipson) [reply arch/x86/boot/compressed/tpm/tpm2_cmds.c:87]: If there are TPM fixes, lets bring them in before the AMD PR goes in. I think that goes for any other general fixes that are found that are not related to adding AMD support.

File: arch/x86/boot/compressed/tpm/tpm2_cmds.c [hunk 5]
```diff
@@ -84,11 +84,13 @@ int tpm2_extend_pcr(struct tpm *t, u32 pcr,
...
+
 	ret = tpm2_alloc_cmd(b, &cmd, TPM_ST_SESSIONS, TPM_CC_PCR_EXTEND);
 	if (ret < 0)
 		goto out;
 
-	cmd.handles = (u32 *)tpmb_put(b, sizeof(u32));
+	cmd.handles = (u32 *)tpmb_put(b, 2*sizeof(u32));
```
Comment (dpsmith) [inline arch/x86/boot/compressed/tpm/tpm2_cmds.c:93]: I think this is left over from the corrections you all made to the previous version which was missing the tpmb_put for the auth handle. That has been corrected so this ends up allocating an extra handle.

File: arch/x86/boot/compressed/tpm/tpm2_cmds.c (hunk 4 above)
Comment (dpsmith) [reply arch/x86/boot/compressed/tpm/tpm2_cmds.c:87]: Yes, this is an improvement to the existing tpm patches, so this should be captured in a delta commit.

File: arch/x86/boot/compressed/sl_main.c [hunk 6]
```diff
@@ -278,30 +284,33 @@ void sl_main(u8 *bootparams)
...
 	/*
 	 * If enable_tpm fails there is no point going on. The entire secure
 	 * environment depends on this and the other TPM operations succeeding.
 	 */
 	tpm = enable_tpm();
-	if (!tpm)
-		sl_txt_reset(SL_ERROR_TPM_INIT);
+	if (!tpm) {
+		if (sl_cpu_type == SL_CPU_INTEL)
+			sl_txt_reset(TXT_SLERROR_TPM_INIT);
```
Comment (dpsmith) [inline arch/x86/boot/compressed/sl_main.c]: Please use the value you replaced, SL_ERROR_TPM_INIT, so that you don't have to undo this in your third commit.

File: arch/x86/boot/compressed/sl_main.c [hunk 7]
```diff
@@ -278,30 +284,33 @@ void sl_main(u8 *bootparams)
...
 
-	/*
-	 * Locality 2 is being opened so that the DRTM PCRs can be updated,
-	 * specifically 17 and 18.
-	 */
-	if (tpm_request_locality(tpm, 2) == TPM_NO_LOCALITY)
-		sl_txt_reset(SL_ERROR_TPM_GET_LOC);
+	if (tpm_request_locality(tpm, 2) == TPM_NO_LOCALITY) {
+		if (sl_cpu_type == SL_CPU_INTEL)
+			sl_txt_reset(TXT_SLERROR_TPM_GET_LOC);
```
Comment (dpsmith) [inline arch/x86/boot/compressed/sl_main.c]: Please use the value you replaced, SL_ERROR_TPM_GET_LOC, so that you don't have to undo this in your third commit.

File: arch/x86/boot/compressed/tpm/tpm2_cmds.c (hunk 4 above)
Comment (krystian-hebel) [reply arch/x86/boot/compressed/tpm/tpm2_cmds.c:87]: This function should probably be renamed, or split into two separate functions. The reason it is called here is that it also resets head/tail/len of that buffer.

File: arch/x86/boot/compressed/sl_main.c (hunk 1 above)
Comment (krystian-hebel) [reply arch/x86/boot/compressed/sl_main.c]: (Un)done

File: arch/x86/boot/compressed/sl_stub.S (hunk 2 above)
Comment (krystian-hebel) [reply arch/x86/boot/compressed/sl_stub.S]: Fixed

File: arch/x86/boot/compressed/sl_main.c (hunk 6 above)
Comment (krystian-hebel) [reply arch/x86/boot/compressed/sl_main.c]: Done

File: arch/x86/boot/compressed/sl_main.c (hunk 7 above)
Comment (krystian-hebel) [reply arch/x86/boot/compressed/sl_main.c]: Done

File: arch/x86/boot/compressed/tpm/tpm2_cmds.c (hunk 5 above)
Comment (krystian-hebel) [reply arch/x86/boot/compressed/tpm/tpm2_cmds.c]: Done

File: arch/x86/boot/compressed/tpm/tpm2_cmds.c (hunk 4 above)
Comment (krystian-hebel) [reply arch/x86/boot/compressed/tpm/tpm2_cmds.c]: Rebased

File: arch/x86/boot/compressed/sl_main.c (hunk 3 above)
Comment (krystian-hebel) [reply arch/x86/boot/compressed/sl_main.c]: Done

File: arch/x86/boot/compressed/sl_stub.S [hunk 8]
```diff
@@ -168,9 +168,25 @@ SYM_FUNC_START(sl_stub)
...
+	cmpl	$(AMD_CPUID_MFGID_ECX), %ecx
+	jnz	.Ldo_unknown_cpu
+
+	/* Know it is AMD */
+	movl	$(SL_CPU_AMD), sl_cpu_type(%ebp)
+
+	/*
+	 * Enable global interrupts including SMI and NMI (GIF).
+	 */
+	stgi
```
Comment (rossphilipson) [inline arch/x86/boot/compressed/sl_stub.S:188]: One final thought. Did you want to move this until after the measurements in sl_main()? I am planning to move the instruction to enable SMI until the end of that too. @dpsmith  was saying we should keep SMI disabled until after the measurements are done.

Comment (krystian-hebel) [reply arch/x86/boot/compressed/sl_stub.S:188]: I am not sure about that. On the one hand it would prevent SMM code from messing up communication with TPM, on the other hand, we shouldn't trust such SMM if it does.

IMHO it would be better to delay this until interrupt handlers are installed, as it enables not only SMI.

Comment (rossphilipson) [reply arch/x86/boot/compressed/sl_stub.S:188]: I agree with you first point on SMM and your second point I think is a larger discussion. We can just leave it there for now if you want to get this in...

Comment (krystian-hebel) [reply arch/x86/boot/compressed/sl_stub.S:188]: I'm not devoted to keeping it here, we can move it. Lets keep it as similar as possible for Intel and AMD.

Comment (rossphilipson) [reply arch/x86/boot/compressed/sl_stub.S:188]: For now what you have is inline with what TXT is doing. So let's just keep it here for now.
